### PR TITLE
Change CosmosClient in TRE API to be asynchronous

### LINKED ISSUE #2599: Change CosmosClient in TRE API to be asynchronous
## Description
As for today, the CosmosClient that is being used in the TRE API is synchronous, in oppose to the ServiceBus client which is asynchronous.
This user story is about refactoring the parts that use the cosmos client and replace it with the asynchronous version.

### PR DESCRIPTION
**Pull request by anatbal**
# Resolves #2599 

## What is being addressed
Converting the CosmosClient we are using in the API from synchronous to asynchronous.
One of the major problems we have in design is the fact that in BaseRepository class we are making a `create_container_if_not_exist` call inside the constructor. This has caused a breaking change in the way we initialize each Repository class and also might not be the right way to implement. 

This PR continues with this approach (no design change), but it might be worth to consider a design change.

## Instructions for reviewer
Although it seems many files had been changed, it mostly because "async" and "await" had to be added in a lot of places.
The important files that actually had been changed are:
1.  The actual client that was changed and the methods that relates to it- [api_app/api/dependencies/database.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-26a69ea7be71328f172325f576a985a9e09973d59cfbad853b3dff44ba187f16)

2. **All db repository classes:** The same change in all of them, beside base repo. Since now the get_container method in base repo is asynchronous, it means all of the repositories constructers are supposed to be async as well. Such thing is not possible, hence I converted all constructures to a "create" factory method which initiates the db client and container.
     a. [api_app/db/repositories/airlock_requests.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-a40fdbce382c18939912136596986ee345f902aa1f0af94be8c2d77db87b116b)
     b. [api_app/db/repositories/base.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-62d1d4d2495ddef7ad529cedf93bd3d960b7c9375ae949f5e11b44c1baf24bee)
     c. [api_app/db/repositories/operations.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-b80128c8331aa2d5071f44eeea16ec2b3552c393cc3e565ac19f631b8b27ec6a)
    d. [api_app/db/repositories/resource_templates.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-95d190fe7ba3c466c36f444dace2cd9cddff37578721220da82f91536668f771)
    e. [api_app/db/repositories/resources.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-c99edc996fc518d7fa4de950a1ac4363ffb2a45a42710e7887b514b4a1df6c10)
    f. [api_app/db/repositories/shared_services.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-b37de085862a3a5620968ffd0978c607d049e2ed559c8e11c8c121136d8c0cd0)
    g. [api_app/db/repositories/user_resources.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-b0ad3afd3a5d645ddb3fb52f590e67a4c948cc9cd7391b152966018062cc1f1f)
    h. [api_app/db/repositories/workspaces.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-730452840b06eb8de3036f7590db080578fa76d212cc7d9224bdb6cfe9e4df13)
    i. [api_app/db/repositories/workspace_services.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-c3039e64a1c13757af4d6ac4a62947c0791cadabce83ca211bef8f3d16653d68)

3.  **All db migrations classes:** The same change as the one above.
    a. [api_app/db/migrations/airlock.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-416a77dbf04c1009a1aca842f8598d6e9a2489cfd97b361cef4476659d0409ce)
    b. [api_app/db/migrations/resources.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-a956ddd1b5d9cbcac134676482addfc44096f6ee25e8a552a3f38689dc7ed3db)
   c. [api_app/db/migrations/workspaces.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-38fccc78c6994da0cf3abb0b52385a3c31a236f5e31854489f57c69ebf842390)
   d. [api_app/db/migrations/shared_services.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-119e2e638b5ec8567313ce37d69389b84a860df657a9e0fcca6627e11139a851)

4. DeploymentStatus class: The same change as above - [api_app/service_bus/deployment_status_updater.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-ccd27a5100d83ebe31fc97f67fecae5f5388506cae377d4cb00bec820e4424c2)

5. Health Check - removed duplicated function - [api_app/services/health_checker.py](https://github.com/microsoft/AzureTRE/compare/main...anatbal:feature/asynccosmosclient?expand=1#diff-996ef01f243a5a73eb5332f2e7d2d51fc4c47309cf3ec5230e72a895ec51f600)

6.  Unit tests - Since both the cosmos client and the initiating of Repository class have been changed, it caused a lot of changes to the unit tests as well.  Most of the changes are using pytest_asyncio package and replacing MagicMocks with AsyncMocks.

### CODE REVIEW

File: api_app/api/dependencies/database.py [hunk 1]
```diff
@@ -1,60 +1,80 @@
...
-        primary_master_key = get_store_key()
+        database_credentials = await get_credential_async()
+        primary_master_key = await get_store_key(database_credentials)
 
         if config.STATE_STORE_SSL_VERIFY:
-            cosmos_client = CosmosClient(url=config.STATE_STORE_ENDPOINT, credential=primary_master_key, consistency_level='Session')
+            cosmos_client = CosmosClient(url=config.STATE_STORE_ENDPOINT, credential=primary_master_key)
         else:
             # ignore TLS (setup is a pain) when using local Cosmos emulator.
-            cosmos_client = CosmosClient(config.STATE_STORE_ENDPOINT, primary_master_key, consistency_level='Session', connection_verify=False)
```
Comment (guybartal) [inline api_app/api/dependencies/database.py:26]: Just wondering, why consistency level setting to `Session` was removed here? 
do we want to rely on the account level consistency setting which is currently `Bounded Staleness`

Comment (anatbal) [reply api_app/api/dependencies/database.py:26]: Great questions, I discussed it offline with @tamirkamara and the reason it was set to Session from the begging is simply because it was the default behavior of the sdk client. However, the async client does not seem to support 'Session', which I had to remove. I went and saw that the account level setting works, but if you have a better suggestion I would love your opinion since I am not very familiar with the subject.

File: api_app/db/repositories/base.py [hunk 2]
```diff
@@ -10,52 +11,56 @@
 
 
 class BaseRepository:
-    def __init__(self, client: CosmosClient, container_name: str = None) -> None:
-        self._client: CosmosClient = client
-        self._container: ContainerProxy = self._get_container(container_name)
+    @classmethod
+    async def create(cls, client: CosmosClient, container_name: str = None):
+        cls = BaseRepository()
```
Comment (guybartal) [inline api_app/db/repositories/base.py:16]: `BaseRepository` in my opinion should have been an abstract class originally, but this change is not related directly to support async calls to cosmos db so maybe we refactor that in a separate issue.
Consider removing this initialization as this Base class shouldn't be instantiated directly.

File: api_app/db/repositories/operations.py [hunk 3]
```diff
@@ -18,8 +18,13 @@
 
 
 class OperationRepository(BaseRepository):
-    def __init__(self, client: CosmosClient):
-        super().__init__(client, config.STATE_STORE_OPERATIONS_CONTAINER)
+    @classmethod
+    async def create(cls, client: CosmosClient):
+        cls = OperationRepository()
+        base_repo = await super().create(client, config.STATE_STORE_OPERATIONS_CONTAINER)
+        cls._container = base_repo._container
```
Comment (guybartal) [inline api_app/db/repositories/operations.py:25]: In case you remove the initialization of the base class in the factory method as commented above,
you should remove those lines as well, and drop the `base_repo` local variable .
```
cls._container = base_repo._container
cls._client = base_repo._client
```
same goes to all repositories.

File: api_app/db/repositories/base.py [hunk 4]
```diff
@@ -10,52 +11,56 @@
...
+        cls._client: CosmosClient = client
+        cls._container: ContainerProxy = await cls._get_container(container_name)
+        return cls
 
     @property
     def container(self) -> ContainerProxy:
         return self._container
 
-    def _get_container(self, container_name) -> ContainerProxy:
+    async def _get_container(self, container_name) -> ContainerProxy:
```
Comment (guybartal) [inline api_app/db/repositories/base.py:25]: this should be `@classmethod` as well

File: api_app/db/repositories/operations.py (hunk 3 above)
Comment (anatbal) [reply api_app/db/repositories/operations.py]: Fixed

File: api_app/db/repositories/base.py (hunk 2 above)
Comment (anatbal) [reply api_app/db/repositories/base.py]: Fixed, and as discussed offline, will open a separate issue

**Review by guybartal — APPROVED** [review]
LGTM, good job!

**Comment by eladiw** [comment]
@anatbal , Nice PR description!
(didn't go over all the files but saw the main changes - looks good)

**Comment by anatbal** [comment]
https://github.com/microsoft/AzureTRE/actions/runs/3741037415 
This run has succeeded, merging the PR
